### DISCUSSION TITLE
Initial setup questions

### DISCUSSION THREAD

**Discussion by embold-given** [SuperGoodSoft/solidus_taxjar#131] [Support]
Initial setup looks pretty straightforward, but I'm not seeing what I would expect.

With everything configured as instructed, checkout doesn't show any taxes.

I'm on Solidus 3 and Ruby 3

I created the TaxRate with the name of "Sales Tax". It wasn't specified, but I added a rate of 0 because it wouldn't let me save without a numeric value.

I created the `taxjar.rb` and have been using that for general configuration settings.

It wasn't in the readme here, but I saw it in the instructions on rubygems, I've tried adding 
```
Spree.config do |config|
  config.tax_calculator_class = SuperGood::SolidusTaxjar::TaxCalculator
end
```
to each of `taxjar.rb`, to `spree.rb` and to `solidus.rb` (but not all at the same time, that's just madness)

We're using encrypted credentials on this project, so I added the `TAXJAR_API_KEY` to the environment variables via my spree.rb file like so:
```ENV['TAXJAR_API_KEY'] = Rails.application.credentials[:development][:taxjar_api_key]```

In TaxJar I have taxes set for the state that I'm using for test orders.

Between rubygems and github I think I've got everytrhing set up the way I'm supposed to, but it's not acting how I expect. Am I missing something?

Thanks!

**Comment by Noah-Silvera** [comment]
Hey @embold-given!

I wanted to just clarify something about your configuration to make sure everything is setup right!

```rb
Spree.config do |config|
  config.tax_calculator_class = SuperGood::SolidusTaxjar::TaxCalculator
end
```

This piece of config will reconfigure the "Default Tax" base calculator option to use the TaxJar calculator. 

* The `TaxRate` you create **has to** use the "Default Tax" calculator as its base calculator, and that `TaxRate` has to apply to the Zone/Tax Category of your order for Solidus to use it.

* The rate you set for the `TaxRate` actually doesn't matter. Solidus has a confusing UI for `TaxRates` where it shows all the options that apply for any possible calculator. So for the "Flat Rate" base calculator, the "rate" option would apply, but for the TaxJar calculator, it actually doesn't matter.

* Having the config option in any of `taxjar.rb`, to `spree.rb` and to `solidus.rb` should work fine, it just needs to be in some initializer!

* The state you configured on the `taxjar.com` also has to be marked with "Nexus: Yes".

If it's still not showing taxes after verifying your configuration, could you search your server logs for `TaxJar response` when trying to checkout an order see if it shows up? Then we can know if the extension is actually hitting the API or not.

I also wanted to note that we currently are not actively testing against Solidus 3 -  it's a high priority task to test against that version and fix any potential problems that arise, but as of now, it's untested. 

---- 

As a side note, thanks for pointing out the missing info the readme! We'll be sure to correct that.

**Reply by embold-given** [reply]
- I can confirm I've got it using the default calculator. I have the TaxRate set to the general "North America" zone, but we only currently have TaxJar set up for 1 state at the moment. I'll set up a zone for just that state and will see if that makes a difference.
- I figured the rate value wasn't important and TaxJar would override whatever was set, was just throwing it out on the off chance it was something.
- I wouldn't have guessed it would have made a difference, but I'm still learning the ins and outs of Ruby / Solidus, so I'm never quite sure what's important and what's not.
- I can confirm that the state in TaxJar is set as a Nexus.

I'll double check all my settings, make sure I didn't accidentally uncheck something while I was poking around and then will check the logs.

I was afraid that Solidus 3 might be the issue. I saw `solidus_core >= 2.4.0` on rubygems and I was hopeful that meant 3 would be alright.

Thanks for the direction, I'll report back with whatever the results are! :D

**Reply by embold-given** [reply]
Double checked all my settings, deleted and created a new "Default" TaxCategory, deleted and created a new "Sales Tax" TaxRate, even set up a new zone that's just the one state I have set up as a Nexus in TaxJar, no references at all in the logs for "TaxJar". I did clear the logs before testing, but I went through the whole process from adding an item to the cart to placing the order, still nothing.

**Reply by jarednorman** [reply]
One gotcha is that the calculator will swallow any exceptions it encounters (as we'd rather not charge tax than break checkout.) It's worth [configuring an error reporter](https://github.com/SuperGoodSoft/solidus_taxjar/blob/master/lib/super_good/solidus_taxjar.rb#L53-L55) or double checking you don't see anything in your logs to see if that uncovers the issue.

**Reply by embold-given** [reply]
Oh good point, the project is new enough that we haven't got to that step yet, so I didn't configure one and assumed that whatever was default would report them. Apparently a bad assumption on my part. I'll do that next and will see if anything turns up. Thanks!

**Reply by embold-given** [reply]
New info! Sentry is so handy. I can't believe I forgot to set it up earlier.

```
Taxjar::Error::Unauthorized
Not authorized for route 'POST /v2/taxes'
```

Looks like it's getting triggered in `Spree::CheckoutController#update` 

I'll pick up the trail there and see if I can track down exactly what's going on.

**Reply by jarednorman** [reply]
That sounds like you're missing TaxJar credentials.

**Reply by embold-given** [reply]
Turns out I was trying to use the sandbox api key but was pointed at the live API URL. I found a reference to the ENV value in `api.rb`. Set that in my config and now I'm getting 
```
Taxjar::Error::BadRequest Spree::CheckoutController#update
No from address, no nexus address, and no address on file
```

**Reply by jarednorman** [reply]
Looks like you've not configured a nexus in TaxJar. TaxJar needs to know where you have nexus (physical presence) in order to calculate taxes.

**Reply by embold-given** [reply]
I have an address set in TaxJar, which it sounds like it should be using if there isn't anything sent through the API. I've also made sure that there's an address set for the stock location but it looks like that's not the right spot, either. Is there another spot I missed for setting the "from" address?

**Reply by embold-given** [reply]
There is an address set up in TaxJar in the Nexus section
![image](https://user-images.githubusercontent.com/65610008/144913264-bcfc111c-e08a-449f-bcef-456b914f663d.png)

Is this not the right place for that?

edit:
Also found this section where it looks like it says it's enabled for the API:
![image](https://user-images.githubusercontent.com/65610008/144913879-99f33bd8-49a8-4d42-aea0-5578e7b6e976.png)

**Reply by embold-given** [reply]
Ok, looks like we'll need to have our client actually go through and set up some more of the nexus info, I see it's set up as a nexus state, but there's all sorts of info that's not in here. I'm guessing that's what's causing problems.

**Reply by embold-given** [reply]
Just for confirmation, I have this address set:
![image](https://user-images.githubusercontent.com/65610008/144919586-79997717-5af6-4914-a3fa-75c757d8dc1f.png)

The wording makes it sound like if there's no address set, it would use this one by default. It's set, so I would think rates should be returning. If that's not the case, I'll take any further support issues to TaxJar and have them help with the next steps.
